Mathlib/Algebra/BigOperators/Expect.lean

```diff
@@ -0,0 +1,400 @@
+
+namespace BigOperators
+open Batteries.ExtendedBinder Lean Meta
+
+/--
+* `𝔼 i ∈ s, f i` is notation for `Finset.expect s f`. It is the expectation of `f i` where `i`
+  ranges over the finite set `s` (either a `Finset` or a `Set` with a `Fintype` instance).
+* `𝔼 i, f i` is notation for `Finset.expect Finset.univ f`. It is the expectation of `f i` where `i`
+  ranges over the finite domain of `f`.
+* `𝔼 i ∈ s with p i, f i` is notation for `Finset.expect (Finset.filter p s) f`.
+* `𝔼 (i ∈ s) (j ∈ t), f i j` is notation for `Finset.expect (s ×ˢ t) (fun ⟨i, j⟩ ↦ f i j)`.
+
+These support destructuring, for example `𝔼 ⟨i, j⟩ ∈ s ×ˢ t, f i j`.
+
+Notation: `"𝔼" bigOpBinders* ("with" term)? "," term` -/
+scoped syntax (name := bigexpect) "𝔼 " bigOpBinders ("with " term)? ", " term:67 : term
+
+scoped macro_rules (kind := bigexpect)
+  | `(𝔼 $bs:bigOpBinders $[with $p?]?, $v) => do
+    let processed ← processBigOpBinders bs
+    let i ← bigOpBindersPattern processed
+    let s ← bigOpBindersProd processed
+    match p? with
+    | some p => `(Finset.expect (Finset.filter (fun $i ↦ $p) $s) (fun $i ↦ $v))
+    | none => `(Finset.expect $s (fun $i ↦ $v))
+
+open Lean Meta Parser.Term PrettyPrinter.Delaborator SubExpr
+open Batteries.ExtendedBinder
+
+/-- Delaborator for `Finset.expect`. The `pp.piBinderTypes` option controls whether
+to show the domain type when the expect is over `Finset.univ`. -/
+@[scoped delab app.Finset.expect] def delabFinsetexpect : Delab := whenPPOption getPPNotation do
+  let #[_, _, _, s, f] := (← getExpr).getAppArgs | failure
+  guard $ f.isLambda
+  let ppDomain ← getPPOption getPPPiBinderTypes
+  let (i, body) ← withAppArg $ withBindingBodyUnusedName fun i => do
+    return (i, ← delab)
+  if s.isAppOfArity ``Finset.univ 2 then
+    let binder ←
+      if ppDomain then
+        let ty ← withNaryArg 0 delab
+        `(bigOpBinder| $(.mk i):ident : $ty)
+      else
+        `(bigOpBinder| $(.mk i):ident)
+    `(𝔼 $binder:bigOpBinder, $body)
+  else
+    let ss ← withNaryArg 3 $ delab
+    `(𝔼 $(.mk i):ident ∈ $ss, $body)
+
+end BigOperators
+
+open scoped BigOperators
+
+namespace Finset
+section AddCommMonoid
+variable [AddCommMonoid M] [Module ℚ≥0 M] [AddCommMonoid N] [Module ℚ≥0 N] {s : Finset ι}
+  {f g : ι → M} {m : N → M}
+
+lemma expect_univ [Fintype ι] : 𝔼 i, f i = (∑ i, f i) /ℚ Fintype.card ι := by
+  rw [expect, card_univ]
+
+@[simp] lemma expect_empty (f : ι → M) : expect ∅ f = 0 := by simp [expect]
+@[simp] lemma expect_singleton (f : ι → M) (i : ι) : expect {i} f = f i := by simp [expect]
+@[simp] lemma expect_const_zero (s : Finset ι) : 𝔼 _i ∈ s, (0 : M) = 0 := by simp [expect]
+
+@[congr]
+lemma expect_congr {t : Finset ι} (hst : s = t) (h : ∀ i ∈ t, f i = g i) :
+    𝔼 i ∈ s, f i = 𝔼 i ∈ t, g i := by rw [expect, expect, sum_congr hst h, hst]
+
+lemma expectWith_congr (p : ι → Prop) [DecidablePred p] (h : ∀ i ∈ s, p i → f i = g i) :
+    𝔼 i ∈ s with p i, f i = 𝔼 i ∈ s with p i, g i :=
```

Comment:
This is more `expectWith_congrFun`, not a fully general congruence, which could be
```lean
lemma expectWith_congr (p q : ι → Prop) [DecidablePred p] [DecidablePred q] (hst : s = t)
    (hpq : ∀ i ∈ s, p i ↔ q i) (h : ∀ i ∈ s, p i → f i = g i) :
    𝔼 i ∈ s with p i, f i = 𝔼 i ∈ t with q i, g i
```